File: web-face/src/app/core/services/device.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { Device } from '../models/device.model';
import { DeviceStatus } from '../models/enums';
import { Page, Pageable } from '../models/page.model';
import { CrudService } from './crud.service';
import { buildHttpParams } from './http-params.util';

export interface DeviceSearchParams extends Pageable {
  q?: string;
  categoryId?: string;
  tags?: string[];
  status?: DeviceStatus;
}

@Injectable({ providedIn: 'root' })
export class DeviceService extends CrudService<Device> {
  constructor() {
    super('devices');
  }

  findByDeviceIdentifier(deviceIdentifier: string): Observable<Device> {
    return this.http.get<Device>(`${this.resourceUrl}/by-identifier/${deviceIdentifier}`);
  }

  /** Recomputes and persists a fresh deviceIdentifier for an existing device. Requires device:write. */
  regenerateIdentifier(id: string): Observable<Device> {
    return this.http.post<Device>(`${this.resourceUrl}/${id}/identifier/regenerate`, {});
  }

  /** Paginated/sortable listing, optionally filtered by free-text query, category, tags, and/or status. */
  search(params: DeviceSearchParams = {}): Observable<Page<Device>> {
    return this.http.get<Page<Device>>(`${this.resourceUrl}/search`, { params: buildHttpParams(params) });
  }
}

File: core/src/main/java/com/csl/lasform/service/DeviceService.java
package com.csl.lasform.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.csl.lasform.model.entity.Device;
import com.csl.lasform.model.entity.enums.DeviceStatus;

public interface DeviceService extends CrudService<Device, String>, ImageAttachable {

    Device findByDeviceIdentifier(String deviceIdentifier);

    /** Paginated/sortable listing, optionally filtered by free-text {@code q} (name/identifier/tags), category, tags, and/or status. */
    Page<Device> search(String q, String categoryId, List<String> tags, DeviceStatus status, Pageable pageable);
}
